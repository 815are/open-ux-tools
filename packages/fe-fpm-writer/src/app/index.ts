import { create as createStorage } from 'mem-fs';
import type { Editor } from 'mem-fs-editor';
import { create } from 'mem-fs-editor';
import { join } from 'path';
import { satisfies } from 'semver';
import type { SAPJSONSchemaForWebApplicationManifestFile as Manifest } from '../common/manifest';

export interface FPMConfig {
    fcl?: boolean;
}

/**
 * Enable the flexible programming model for an application.
 *
 * @param {string} basePath - the base path
 * @param {CustomTableColumn} config - the custom column configuration
 * @param {Editor} [fs] - the mem-fs editor instance
 * @returns {Promise<Editor>} the updated mem-fs editor instance
 */
export function enableFPM(basePath: string, config: FPMConfig = {}, fs?: Editor): Editor {
    if (!fs) {
        fs = create(createStorage());
    }

    const manifestPath = join(basePath, 'webapp/manifest.json');
    const manifest = fs.readJSON(manifestPath) as any as Manifest;

    // add FE libs is not yet added
    if (!manifest['sap.ui5']?.dependencies?.libs?.['sap.fe.templates']) {
        fs.extendJSON(manifestPath, {
            'sap.ui5': {
                dependencies: {
                    libs: {
                        'sap.fe.templates': {}
                    }
                }
            }
        });
    }

    // if a minUI5Version is set and it is smaller than the minimum required, increase it
    if (
        manifest['sap.ui5']?.dependencies.minUI5Version &&
        !satisfies(manifest['sap.ui5']?.dependencies.minUI5Version, '>=1.86')
    ) {
        fs.extendJSON(manifestPath, {
            'sap.ui5': {
                dependencies: {
                    minUI5Version: '1.86.0'
                }
            }
        });
    }
    // enable FCL if requested
    if (config.fcl) {
        fs.extendJSON(manifestPath, {
            'sap.ui5': {
                rootView: {
                    viewName: 'sap.fe.templates.RootContainer.view.Fcl',
                    type: 'XML',
                    async: true,
                    id: 'appRootView'
                },
                routing: {
                    config: {
                        routerClass: 'sap.f.routing.Router'
                    }
                }
            }
        });
    }

    // replace Component.js
    const componentTemplate = join(__dirname, '../../templates/app/Component.js');
    fs.copyTpl(componentTemplate, join(basePath, 'webapp/Component.js'), manifest['sap.app']);

    return fs;
}
