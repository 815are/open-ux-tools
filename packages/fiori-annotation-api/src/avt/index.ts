export { convertAnnotationFile } from './annotations';
export { convertMetadataToAvtSchema } from './metadata';
