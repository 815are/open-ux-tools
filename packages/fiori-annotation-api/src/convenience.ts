import type {
    RawSchema,
    RawEntityContainer,
    RawEntitySet,
    RawEntityType,
    RawComplexType,
    RawAction,
    RawProperty,
    RawV4NavigationProperty,
    AnnotationList,
    ActionParameter
} from '@sap-ux/vocabularies-types';
import type { MetadataService, MetadataElement, Path } from '@sap-ux/odata-metadata';

/**
 * convert metadata to AVT Schema representation
 * @param metadataService
 */
export function convertMetaDataToAvtSchema(metadataService: MetadataService): RawSchema {
    const rootLevelElement = metadataService.getRootMetadataElements();
    // Schema representation includes entityTypes, entitySets, associations (not supported here), a single entityContainer and actions
    // collect these objects based on the (multiple) EdmTargetKinds of each metadata element
    const entityTypes: RawEntityType[] = [];
    const complexTypes: RawComplexType[] = [];
    const entitySets: RawEntitySet[] = [];
    let entityContainer: RawEntityContainer = {
        _type: 'EntityContainer',
        fullyQualifiedName: ''
    };
    const actions: RawAction[] = [];
    rootLevelElement.forEach((element: MetadataElement, elementKey: Path) => {
        const targetKinds = metadataService.getEdmTargetKinds(element.path);
        if (targetKinds.includes('EntityContainer')) {
            entityContainer = {
                _type: 'EntityContainer',
                name: element.name,
                fullyQualifiedName: elementKey
            };
            // entity sets as children of entity container
            for (const subElement of element.content ?? []) {
                const subElementTargetKinds = metadataService.getEdmTargetKinds(subElement.path);
                if (subElementTargetKinds.includes('EntitySet')) {
                    // only EntitySets are considered in AVT Schema representation (no Singleton, Action/FunctionImport etc...)
                    const entitySet: RawEntitySet = {
                        _type: 'EntitySet',
                        name: subElement.name,
                        entityTypeName: subElement.structuredType ?? '', // TODO: check if this causes issues
                        navigationPropertyBinding: {},
                        fullyQualifiedName: subElement.path
                    };
                    entitySets.push(entitySet);
                }
                if (metadataService.ODataVersion === '2.0' && subElementTargetKinds.includes('FunctionImport')) {
                    // remark: AVT fills it differently in parseFunctionImport() - but then annotations are not picked up in convertTypes()
                    const action: RawAction = {
                        _type: 'Action',
                        name: subElement.name,
                        fullyQualifiedName: subElement.path,
                        parameters: [],
                        isBound: false, // always set to false in AVT!
                        sourceType: '', // not available here TODO: check if this causes issues
                        returnType: '', // TODO: check if this causes issues
                        isFunction: false
                    };
                    for (const parameterMdElement of subElement.content ?? []) {
                        if (parameterMdElement) {
                            const parameterTargetKinds = metadataService.getEdmTargetKinds(parameterMdElement.path);
                            if (parameterTargetKinds.includes('Parameter')) {
                                // Set action.sourceType to the type from the parameter '_it'
                                if (parameterMdElement.name === '_it') {
                                    action.sourceType = parameterMdElement.structuredType ?? ''; // TODO: check if this causes issues
                                }
                                const parameter: ActionParameter = {
                                    _type: 'ActionParameter',
                                    isEntitySet: false,
                                    fullyQualifiedName: parameterMdElement.path,
                                    annotations: {},
                                    name: parameterMdElement.name,
                                    type: parameterMdElement.structuredType ?? parameterMdElement.edmPrimitiveType ?? '' // TODO: check if this causes issues
                                };
                                action.parameters.push(parameter);
                            } else if (parameterTargetKinds.includes('ReturnType')) {
                                action.returnType =
                                    parameterMdElement.structuredType ?? parameterMdElement.edmPrimitiveType ?? ''; // TODO: check if this causes issues
                            }
                        }
                    }
                    actions.push(action);
                }
            }
        }
        if (targetKinds.includes('EntityType')) {
            const entityProperties: RawProperty[] = [];
            const entityNavProps: RawV4NavigationProperty[] = [];
            const keys: RawProperty[] = [];
            for (const subElement of element.content ?? []) {
                const subElementTargetKinds = metadataService.getEdmTargetKinds(subElement.path);
                if (subElementTargetKinds.includes('Property')) {
                    const property: RawProperty = {
                        _type: 'Property',
                        name: subElement.name,
                        type: subElement.edmPrimitiveType ?? subElement.structuredType ?? '', // TODO: check if this causes issues
                        fullyQualifiedName: subElement.path
                    };
                    entityProperties.push(property);
                    if (element.keys && element.keys.includes(subElement.name)) {
                        keys.push(property);
                    }
                }
                if (subElementTargetKinds.includes('NavigationProperty')) {
                    const navProp: RawV4NavigationProperty = {
                        _type: 'NavigationProperty',
                        name: subElement.name,
                        fullyQualifiedName: subElement.path,
                        targetTypeName: subElement.structuredType ?? '', // TODO: check if this causes issues
                        isCollection: false,
                        containsTarget: false,
                        partner: '', // TODO: check if this causes issues
                        referentialConstraint: []
                    };
                    entityNavProps.push(navProp);
                }
            }
            const entityType: RawEntityType = {
                _type: 'EntityType',
                name: element.name,
                fullyQualifiedName: elementKey,
                entityProperties: entityProperties,
                keys: keys,
                navigationProperties: entityNavProps,
                actions: {}
            };
            entityTypes.push(entityType);
        }
        if (targetKinds.includes('ComplexType')) {
            const complexTypeProperties: RawProperty[] = [];
            const complexTypeNavProperties: RawV4NavigationProperty[] = [];
            for (const subElement of element.content ?? []) {
                const subElementTargetKinds = metadataService.getEdmTargetKinds(subElement.path);
                if (subElementTargetKinds.includes('Property')) {
                    const property: RawProperty = {
                        _type: 'Property',
                        name: subElement.name,
                        type: subElement.edmPrimitiveType ?? subElement.structuredType ?? '', // TODO: check if this causes issues
                        fullyQualifiedName: subElement.path
                    };
                    complexTypeProperties.push(property);
                }
                if (subElementTargetKinds.includes('NavigationProperty')) {
                    const navProp: RawV4NavigationProperty = {
                        _type: 'NavigationProperty',
                        name: subElement.name,
                        fullyQualifiedName: subElement.path,
                        targetTypeName: subElement.structuredType ?? '', // TODO: check if this causes issues
                        isCollection: false,
                        containsTarget: false,
                        partner: '', // TODO: check if this causes issues
                        referentialConstraint: []
                    };
                    complexTypeNavProperties.push(navProp);
                }
            }
            const complexType: RawComplexType = {
                _type: 'ComplexType',
                name: element.name,
                fullyQualifiedName: elementKey,
                properties: complexTypeProperties,
                navigationProperties: complexTypeNavProperties
            };
            complexTypes.push(complexType);
        }
        if (targetKinds.includes('EntitySet')) {
            // for CDS no entity container is present: target kind EntitySet can appear at root level
            const entitySet: RawEntitySet = {
                _type: 'EntitySet',
                name: element.name,
                entityTypeName: element.structuredType || (targetKinds.includes('EntityType') ? element.name : ''),
                navigationPropertyBinding: {},
                fullyQualifiedName: element.path
            };
            entitySets.push(entitySet);
        }
        if (targetKinds.includes('Action') || targetKinds.includes('Function')) {
            // see AVT parseActions()
            const action: RawAction = {
                _type: 'Action',
                name: element.name,
                fullyQualifiedName: element.path,
                parameters: [],
                isBound: false, // TODO: check if this causes issues
                sourceType: '', // not available here // TODO: check if this causes issues
                returnType: '', // TODO: check if this causes issues
                isFunction: false
            };
            for (const subElement of element.content ?? []) {
                const subElementTargetKinds = metadataService.getEdmTargetKinds(subElement.path);
                if (subElementTargetKinds.includes('Parameter')) {
                    const parameter = {
                        _type: 'ActionParameter',
                        isEntitySet: false,
                        fullyQualifiedName: subElement.path,
                        type: subElement.structuredType || subElement.edmPrimitiveType
                    };
                    action.parameters.push(parameter as any);
                } else if (subElementTargetKinds.includes('ReturnType')) {
                    action.returnType = subElement.structuredType ?? subElement.edmPrimitiveType ?? ''; // TODO: check if this causes issues
                }
            }
            actions.push(action);
        }
    });

    const [namespace] = [...metadataService.getNamespaces()];
    if (entityContainer.fullyQualifiedName && namespace) {
        // for CDS no entity container is present: build dummy entity container
        const entityContainerName = 'dummyEntityContainer';
        entityContainer = {
            _type: 'EntityContainer',
            name: entityContainerName,
            fullyQualifiedName: namespace ? namespace + '.' + entityContainerName : entityContainerName
        };
    }

    const annotations: Record<string, AnnotationList[]> = {};
    return {
        namespace,
        annotations,
        entitySets,
        complexTypes,
        entityContainer,
        actions,
        entityTypes,
        associations: [],
        singletons: [],
        associationSets: [],
        typeDefinitions: []
    };
}
