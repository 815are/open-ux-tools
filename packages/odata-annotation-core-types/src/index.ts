export * from './annotation-file';
export * from './base';
export * from './edm';
export * from './edmx';

export { Location, Range, Position, Diagnostic } from './language-server';
