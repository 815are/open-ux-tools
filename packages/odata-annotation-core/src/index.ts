export * from '@sap-ux/odata-annotation-core-types';
export {
    ParsedName,
    ParsedActionFunctionSignature,
    ParsedCollectionIdentifier,
    ParsedIdentifier,
    parseIdentifier,
    toFullyQualifiedName
} from './names';

export {
    NavigationPropertyAnnotationSegment,
    ParsedPath,
    ParsedPathSegment,
    TermCastSegment,
    parsePath,
    toFullyQualifiedPath
} from './paths';
