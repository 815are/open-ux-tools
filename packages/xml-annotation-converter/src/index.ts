export { convertDocument, convertMetadataDocument } from './parser';
