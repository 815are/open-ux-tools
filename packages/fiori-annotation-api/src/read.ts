import type { AnnotationList, ConvertedMetadata, RawMetadata } from '@sap-ux/vocabularies-types';
import { convert } from '@sap-ux/annotation-converter';

import { MetadataService } from '@sap-ux/odata-metadata';

import type { CompiledService, Service } from './services';
import { readXmlAnnotations } from './xml';
import { convertMetadataToAvtSchema, convertAnnotationFile } from './avt';

/**
 * Reads annotations for a specific service in an application.
 *
 * @param service
 * @returns
 */
export function readAnnotations(service: Service): ConvertedMetadata {
    const compiledService = compileService(service);

    const metadataService = new MetadataService();
    metadataService.import(compiledService.metadata);

    const rawMetadata: RawMetadata = {
        version: '2.0',
        identification: 'metadataFile',
        schema: convertMetadataToAvtSchema(metadataService),
        references: []
    };

    const annotationIds = new Set<string>();

    // Merge
    for (const annotationFile of [...compiledService.annotationFiles].reverse()) {
        const targets = convertAnnotationFile(annotationFile);
        const convertedTargets: AnnotationList[] = [];
        for (const target of targets) {
            const convertedTarget: AnnotationList = {
                target: target.target,
                annotations: []
            };
            for (const annotation of target.annotations) {
                const qualifier = annotation.qualifier ? '#' + annotation.qualifier : '';
                const id = `${target.target}@${annotation.term}${qualifier}`;
                if (!annotationIds.has(id)) {
                    annotationIds.add(id);
                    convertedTarget.annotations.push(annotation);
                }
            }
            if (convertedTarget.annotations.length) {
                convertedTargets.push(convertedTarget);
            }
        }
        rawMetadata.schema.annotations[annotationFile.uri] = convertedTargets;
    }

    return convert(rawMetadata);
}

function compileService(service: Service): CompiledService {
    if (service.type === 'local') {
        return readXmlAnnotations(service);
    } else {
        throw new Error(`Unsupported service type "${service.type}"!`);
    }
}
