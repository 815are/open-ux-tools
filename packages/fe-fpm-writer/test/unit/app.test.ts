import { create as createStorage } from 'mem-fs';
import { create, Editor } from 'mem-fs-editor';
import { join } from 'path';
import { enableFPM, MIN_VERSION } from '../../src/app';
import type { Manifest } from '../../src/common/types';

function getTestManifest(settings?: { minVersion?: string }) {
    const manifest: Partial<Manifest> = {
        'sap.app': {
            id: 'my.test.App'
        } as any,
        'sap.ui5': {
            dependencies: {
                minUI5Version: '',
                libs: {}
            },
            routing: {
                routes: [],
                targets: {}
            }
        } as any
    };
    if (settings?.minVersion) {
        manifest['sap.ui5']!.dependencies!.minUI5Version = settings.minVersion;
    }
    return manifest;
}

describe('CustomApp', () => {
    const testDir = '' + Date.now();
    let fs: Editor;

    beforeEach(() => {
        fs = create(createStorage());
        fs.delete(testDir);
    });

    describe('enableFPM', () => {
        test('valid app with minimum version', async () => {
            const target = join(testDir, 'minimal-input');
            fs.writeJSON(join(target, 'webapp/manifest.json'), getTestManifest());
            await enableFPM(target, {}, fs);
            expect(fs.readJSON(join(target, 'webapp/manifest.json'))).toMatchSnapshot();
        });

        test('valid app with a too low minimum version', async () => {
            const target = join(testDir, 'minimal-input');
            fs.writeJSON(join(target, 'webapp/manifest.json'), getTestManifest({ minVersion: '1.23.4' }));
            await enableFPM(target, {}, fs);
            const manifest = fs.readJSON(join(target, 'webapp/manifest.json')) as Manifest;
            expect(manifest['sap.ui5']?.dependencies?.minUI5Version).toBe(MIN_VERSION);
        });
    });
});
