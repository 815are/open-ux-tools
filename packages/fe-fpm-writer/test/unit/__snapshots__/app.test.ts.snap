// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomApp enableFPM valid app with minimum version 1`] = `
Object {
  "sap.app": Object {
    "id": "my.test.App",
  },
  "sap.ui5": Object {
    "dependencies": Object {
      "libs": Object {
        "sap.fe.templates": Object {},
      },
      "minUI5Version": "",
    },
    "routing": Object {
      "routes": Array [],
      "targets": Object {},
    },
  },
}
`;
